dali/pipeline/data/types.h

```diff
@@ -318,7 +319,8 @@ DALI_REGISTER_TYPE(uint8, DALI_UINT8);
 DALI_REGISTER_TYPE(int16, DALI_INT16);
 DALI_REGISTER_TYPE(int32, DALI_INT32);
 DALI_REGISTER_TYPE(int64, DALI_INT64);
-DALI_REGISTER_TYPE(float16, DALI_FLOAT16);
+DALI_REGISTER_TYPE(float16, DALI_FLOAT16);   // GPU implementation of half
+DALI_REGISTER_TYPE(_float16, DALI_FLOAT16);  // CPU implementation of half
```

Comment:
I am not sure if registering two types for the same enum is a good idea, as for both types we will register the same TypeInfo object in TypeTable::type_info_map_ (the second call to TypeTable::RegisterType() will overwrite the first one). 
Due to this, I am not sure if some kind of common placeholder type for float16 is not needed, or maybe we need to split it to DALI_FLOAT16_CPU and DALI_FLOAT16_GPU for now.

Reply:
Actually, nothing wrong with two types sharing the same ID, unless the real data behind these types are NOT different (and, definitely, it's not the case for `float16/_float16`). 

Currently these IDs are used in
(1) Tensors/TensorLists;
(2) in macros like `DALI_TYPE_SWITCH / DALI_IMAGE_TYPE_SWITCH` etc.
(3) description of the operators

As I could see, their usage (1) is not quite significant. For instance, after implementation of CPU version for "half" all tests are fine even when `half_float::half` OR `uint16_t` (which both are `_float16` today) were NOT registered at all. In that case new types are registered "on-the-fly" with the IDs > DALI_DATATYPE_END (1001, 1002, etc)

Case (2) is much more important mostly because 
(a) we are using templated methods here;
(b) these templates are used in `*.cc` and `*.cu` files;
(c) some templates for CPU and GPU (and templates for "half" are among them) are not compatible.

Because of that it is not convenient to split DALI_FLOAT16 into DALI_FLOAT16_CPU / DALI_FLOAT16_GPU - we will need to have also two classes of  DALI_TYPE_SWITCH macros (separately for CPU/GPU).

This split is also inconvenient for the usage of these ID's in the description of operators (3) like the following one:
```
  pipe.AddOperator(
      OpSpec("CropMirrorNormalize")
      .AddArg("device", "gpu")
      .AddArg("output_type", DALI_FLOAT16)
      .AddArg("crop", vector<int>{224, 224})
      .AddArg("image_type", img_type)
      .AddArg("mean", vector<float>{128, 128, 128})
      .AddArg("std", vector<float>{1, 1, 1})
      .AddInput("images", "cpu")
      .AddArgumentInput("crop_pos_x", "uniform1")
      .AddArgumentInput("crop_pos_y", "uniform2")
      .AddArgumentInput("mirror", "mirror")
      .AddOutput("final", "gpu"));
```
In that case we will have additional dependencies between
(a) "output_type" which will be DALI_FLOAT16_GPU and `AddOutput("final", "gpu")`
(which is probably OK)
(b) format of images described in `.AddInput("images", "cpu")` and they could be DALI_FLOAT16_CPU.